### PR TITLE
Maintain control of UI thread during Progression navigation

### PR DESCRIPTION
**Pull request by dpoeschl**
Fixes internal TFS bug #1128063

When NavigateTo is invoked on an IGraphNavigateToItem from the UI
thread, we are now careful to maintain control of the UI thread to
ensure any existing NewDocumentStateScope is preserved.

Tagging for CR: @Pilchie @jasonmalinowski @brettfo @rchande @balajikris 
Special Thanks: @matthejo

### CODE REVIEW

**Comment by dpoeschl** [comment]
@brettfo - This is your bug that we talked about last week. Turns out I introduced the regression. :frown:

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphNavigatorExtension.cs [hunk 1]
```diff
@@ -60,10 +60,23 @@ public void NavigateTo(GraphObject graphObject)
...
                     }
 
-                    Task.Factory.SafeStartNew(
-                        () => NavigateOnForegroundThread(sourceLocation, symbolId, project, document),
-                        CancellationToken.None,
-                        ForegroundTaskScheduler);
+                    if (IsForeground())
+                    {
+                        // If we are already on the UI thread, invoke NavigateOnForegroundThread
+                        // directly to preserve any exsting NewDocumentStateScope.
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphNavigatorExtension.cs]: *existing

**Comment by Pilchie** [comment]
:+1:

**Comment by jmarolf** [comment]
:+1:

**Comment by brettfo** [comment]
:shipit:

**Comment by jasonmalinowski** [comment]
:+1:

**Pull request merged**